### PR TITLE
Adding On-demand Training Data Notebook

### PR DESCRIPTION
**Pull request by KennSmithDS**
PR to merge the notebook tutorial for creating on-demand training data from the Planetary Computer data catalog when starting from a Radiant MLHub dataset.

### CODE REVIEW

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 1]
```diff
@@ -0,0 +1,1281 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ccad2439-109f-4eef-ac3b-dac65f54e3aa",
+   "metadata": {},
+   "source": [
+    "Once you have your API key, you will need to create a default profile by setting up a .mlhub/profiles file in your home directory. You can use the `mlhub configure` command line tool to do this:\n",
+    "\n",
+    "`$ mlhub configure`<br>\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:156]:
```
!mlhub configure --api-key={MLHUB_API_KEY}
```

as a regular code cell should work.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 2]
```diff
@@ -0,0 +1,1281 @@
...
+   "source": [
+    "### Fetch label items from BigEarthNet over Luxembourg"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2ba7d7b2-7fed-4412-8c9b-448baad6e595",
+   "metadata": {},
+   "source": [
+    "This helper function below encapsulates the process of querying a STAC API endpoint to fetch an ItemCollection matching query criteria."
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:202]: IMO this helper function isn't adding much value over just using `catalog.search` directly. I'd rather teach users how to use `catalog.search`.

Can you remove uses of `search_stac_api` and the `client_catalog.search` instead?

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 3]
```diff
@@ -0,0 +1,1281 @@
...
+   "source": [
+    "### Download the entire label collection for BigEarthNet from Radiant MLHub"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2e131160-50bb-487f-8fcb-96d37ce80167",
+   "metadata": {},
+   "source": [
+    "We could certainly use the method above to query label Items directly from our connection to the Radiant MLHub API endpoint. However, on very large collections, such as in the case with BigEarthNet, pagination becomes a bottleneck issue in obtaining and resolving STAC items, as it only returns 100 items at a time.  Querying the entire Collection of nearly ~600,000 Items could take hours.\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:341]: FWIW, the `limit` argument in pystac-client controls the size of the pages. But agreed that fetching 600,000 items through an API isn't what we should recommend.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 4]
```diff
@@ -0,0 +1,1281 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "81f5a8c4-c604-4f8a-8fb9-1d90607206ee",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "if not os.path.exists(label_collection_path):\n",
+    "    collection = Collection.fetch(BIGEARTHNET_LABEL_COLLECTION)\n",
+    "    archive_path = collection.download(TMP_DIR)\n",
+    "    !tar -xf {archive_path.as_posix()} -C {TMP_DIR}\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:376]: It's unfortunate that this decompression takes so long :/ Any thoughts on if you can operate directly on the compressed `.gz`  file? Probably not.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 5]
```diff
@@ -0,0 +1,1281 @@
...
+   "source": [
+    "### Fetch source items for random sample from BigEarthNet"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e419d96f-f8b9-4911-b5d1-4a4077767062",
+   "metadata": {},
+   "source": [
+    "If we had the source collection archive downloaded and uncompressed in the same parent directory as the labels collection, we could reference the source Items and images directly. However the BigEarthNet source collection is over 60GB when compressed. Therefore to work around the disk size limitations of a Planetary Computer instance, we can query the same source items from the MLHub API endpoint, the same way we got the labels, but filter to the exact source item using IDs."
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:562]: What's your preferred workflow as a user, and from MLHub's point of view? Do you want people making local copies of this dataset, or do you want them fetching from your storage on demand?

IMO, the ideal workflow is on-demand fetching from blob storage in the same region + caching, but I'm curious what you think.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 6]
```diff
@@ -0,0 +1,1281 @@
...
+    "        Tuple of the BigEarthNet source Item and the Landsat 8 match Item\n",
+    "    \"\"\"\n",
+    "    # get the matching source Item properties\n",
+    "    source_items = search_stac_api(\n",
+    "        catalog_client=mlhub_catalog,\n",
+    "        collections=[BIGEARTHNET_SOURCE_COLLECTION],\n",
+    "        ids=get_source_item_ids(label_item),\n",
+    "    )\n",
+    "\n",
+    "    if source_items:\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:761]: Is there any reason for this if / elif? Are there cases where you don't have a match?

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 7]
```diff
@@ -0,0 +1,1281 @@
...
+    "    Args:\n",
+    "        asset: Asset object from an Item\n",
+    "\n",
+    "    Returns:\n",
+    "        string response URL direct to Asset\n",
+    "    \"\"\"\n",
+    "    response = requests.get(asset.href, allow_redirects=True)\n",
+    "    if response.status_code == 200:\n",
+    "        return response.url\n",
+    "    return None"
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:898]: Do you ever hit this?

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 8]
```diff
@@ -0,0 +1,1281 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "53556cf0-375b-4044-9262-d16de707a13d",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "if best_l8_match:\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:853]: Same question about potentially not having a match.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 9]
```diff
@@ -0,0 +1,1281 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3f3b5d88-7f3c-4e74-8e68-2408ba762b0f",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "explore_search_extent(ItemCollection([best_l8_match]))"
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:868]: Can you also plot the s2 chip's bound here?

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 10]
```diff
@@ -0,0 +1,1281 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "29531759-6d19-4010-8401-eb947a32c515",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "client = dd_client(\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:1034]: Just use `dask.distributed.Client()` (or a GatewayCluster & get_client if doing this on a distributed cluster)

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 11]
```diff
@@ -0,0 +1,1281 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "927a0496-a6c0-4aa2-9181-708904a5bb4f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# this cell will only work on PC or a machine with gateway cluster configured\n",
+    "# gateway = dask_gateway.Gateway()\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:1048]: IMO: pick whether or not you want this example to run on a cluster. If it runs without a cluster, then we can just remove this.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb [hunk 12]
```diff
@@ -0,0 +1,1281 @@
...
+    "    \"\"\"Creates a Landsat 8 chip from BigEarthNet label chip.\n",
+    "\n",
+    "    Args:\n",
+    "        item_path: string path to the label item on disk\n",
+    "\n",
+    "    Returns:\n",
+    "        Landsat 8 DataArray that has been cropped to label bbox\n",
+    "    \"\"\"\n",
+    "    # read label Item object\n",
+    "    label_item = Item.from_file(\n",
```
Comment (TomAugspurger) [inline tutorials/radiant-mlhub-on-demand-training-data.ipynb:1086]: This, and the other functions touch the local filesystem won't work with a distributed cluster. They'd only work with a LocalCluster.

If we're mentioning possibly using a distributed cluster then you'd need to restructure this. Most likely, you'd need to store everything to Azure Blob Storage and use it as a kind of shared file system that each worker could read / write to.

**Review by TomAugspurger — COMMENTED** [review]
Thanks, I left a few comments inline. Still making my way through the example.

One general comment: I'm pretty uncomfortable having moderately complex code in this examples. I'd much prefer that things like `temporal_buffer`, `mind_cloud_cover_scene` and even `get_landsat_8_match` be generalized and put into a dedicated library, where it can be properly unit tested. With it here in a notebook it's not easy to test and not easy to reuse.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 4 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:376]: I agree, it is unfortunate how long the decompression takes.

I didn't have a great work around to this, as it is related to your previous comment about using `limit`, and taking a random sample of Items from the larger dataset of 600,000 Items. 

I suppose if we fetch only a few thousand to begin with, pagination shouldn't be an issue then we don't need to deal with the `.tar.gz` file. However that likely won't be a random result, I'm guessing the API will just return the first XXX Items?

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 7 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:898]: Similar to my comments on `get_landsat_8_match` above, would you prefer that this be a more simplified notebook cell that stores the non-redirected Asset url in a variable, and not check the response status (e.g. leaving it to users and default error codes built into requests library)?

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 2 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:202]: Done

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 9 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:868]: Similar to other comment threads, would it be best to strip this out of the helper function, and just have a few cells of repeated code for exploring the API search results?

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 12 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:1086]: For purposes of this tutorial, seems to make sense to focus on the workflow of gathering and processing the Landsat 8 data off of Planetary Computer with a `LocalCluster`, and not convoluting that with performing this workflow on a distributed cluster.

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 11 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:1048]: Removed

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 10 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:1034]: for both `ps_client` and `dd_client` I simplified by removing the import alias, so users will directly use `distributed.Client` and `pystac_client.Client`

File: tutorials/radiant-mlhub-on-demand-training-data.ipynb (hunk 9 above)
Comment (KennSmithDS) [reply tutorials/radiant-mlhub-on-demand-training-data.ipynb:868]: Clarifying question on this, do you mean the overall bounds of the ItemCollection returned (e.g. minX, minY, maxX, maxY)? Each Item returned from the API search will have it's own bounding box.

**Comment by KennSmithDS** [comment]
@TomAugspurger do you know why the `stackstac.stack` function will add a buffer in the ndarrays returned? Does it have to do with how it reprojects the image data when it's cached?

For example in [this block of code](https://github.com/RadiantMLHub/PlanetaryComputerExamples/blob/911c455cde28c06972637d488935bdb056356f2a/tutorials/radiant-mlhub-on-demand-training-data.ipynb?short_path=ba0f89d#L908), I'm fetching the Sentinel-2 source imagery from the Azure Blob storage for our MLHub. We know the chips all to be 120x120 pixels, but the stack object dimensions vary from 122x122 up to 130x130.

`s2_stack = stack(
    items=ItemCollection([source_item]),
    assets=BIGEARTHNET_RGB_BANDS,
    epsg=rio.open(get_redirect_url(source_item.assets["B02"])).crs.to_epsg(),
    resolution=10,
)`

P.S. sorry I don't know how you're doing the cool Jupyter Notebook integration.

**Comment by KennSmithDS** [comment]
Another issue I was running into working with the Planetary Computer API comes up maybe 25-50% of the time for Landsat 8 data, depending on the scene. I am uncertain how to account for this when fetching the source imagery using `stackstac`. Any suggestions?

`RuntimeError: Error opening 'https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2018/204/033/LC08_L2SP_204033_20180210_20200902_02_T1/LC08_L2SP_204033_20180210_20200902_02_T1_SR_B3.TIF?st=2022-05-10T20%3A23%3A01Z&se=2022-05-11T21%3A08%3A01Z&sp=rl&sv=2020-06-12&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2022-05-11T03%3A19%3A09Z&ske=2022-05-18T03%3A19%3A09Z&sks=b&skv=2020-06-12&sig=/0r1Hy47dEjm4mpzGS5cR6xK9RkgHCeClZjpdmLiGiQ%3D': RasterioIOError('HTTP response code: 403')`

**Comment by KennSmithDS** [comment]
Closing in favor of #171 due to rebasing issue

**Pull request closed without merging**